### PR TITLE
[ie/youtube] Add `po_token`, `visitor_data` and `data_sync_id` extractor-args

### PR DESCRIPTION
**Pull request by seproDev**
### Description of your *pull request* and other information

Adds extractor arguments to pass a poToken (Proof of origin token) generated in a browser. Also added a visitor_data argument for testing purposes, though this should be provided using cookies ( VISITOR_INFO1_LIVE). 

These tokens are required for some clients (such as web), and possibly more in the future.

There have been reports supplying a po token can help avoid getting blocked, though only to an extent. If you are showing abusive/bot activity then still expect to get blocked (slow down).

#### Usage:

**Passing poToken generated by BotGuard (web, web_embedded, etc.):**
```
--cookies cookies.txt --extractor-arg "youtube:player_client=web;po_token=web+M..."
```
Where the cookies are from the session the poToken was generated in.

**Passing poToken generated by DroidGuard (android):**
```
--extractor-arg "youtube:player_client=android;po_token=android+M...;visitor_data=C...;player_skip=webpage,configs"
```
(This also works for web/web_embedded, but you must skip the webpage and cannot pass cookies unless the visitor ID is the same)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
</details>

### CODE REVIEW

**Comment by BigHeadDev** [comment]
Hello, thank you for your fix, but for the sake of convenience, I don't think we should use external scripts to obtain visitorData and poToken. I found that these two values can be obtained from the browser console:

**visitorData**:document._activeElement.__data.data.response.responseContext.webResponseContextExtensionData.ytConfigData.visitorData

**poToken**:ytPubsubPubsubInstance.j[33].player.app.mediaElement.ea.W.Tf

In addition, due to confusion with YouTube's JS, this path may change after the next update on YouTube. I wonder if the yt-dlp team has any idea of applying this solution to the latest version instead of using external scripts to obtain it

image

**Comment by seproDev** [comment]
yt-dlp does not have access to the JavaScript running in your browser.
As these values are part of the browser integrity checks YouTube runs, it is also unfeasible for us to try to generate them ourselves by executing the JS.
Potentially, a plugin could be written that invokes the tool linked above and refreshes the poToken/visitordata when needed, but this is not something for core code imo.

**Comment by BigHeadDev** [comment]
> yt-dlp does not have access to the JavaScript running in your browser. As these values are part of the browser integrity checks YouTube runs, it is also unfeasible for us to try to generate them ourselves by executing the JS. Potentially, a plugin could be written that invokes the tool linked above and refreshes the poToken/visitordata when needed, but this is not something for core code imo.

Thanks for your reply~~~~😄

**Comment by Elite** [comment]
@seproDev Do we need to get this token once or on every request?

**Comment by jzzbeyond** [comment]
I used the generator but got this error below:

yt-dlp --retries 80 "https://www.youtube.com/watch?v=wsPub08_F4k" -f "bv[height<=720]*[ext=mp4]/b[ext=mp4]/bv*/b" -o "uploads/wsPub08_F4k/v0.%(ext)s" --extractor-arg "youtube:player_client=web;visitor_data=CgtXT1pDaTVXRDdtNCjgqNG1BjIKCgJVUxIEGgAgEg%3D%3D;po_token=MnfYd2uTMkQyC9hMzCCjgXgliR1QlPFfIEPzJ9quodnHPW1e-wkdYyTyy8odu5Yonttw65xl6Z9crU1eT-pwUkhXbW9e9Wkkbw_vrDWxy1IIsjBPI8NeJtOhMXLmWUklMBDEOPGNGgiY1Qs01-axIzh1fPvq2CwGXQ==player_skip=webpage."
[youtube] Extracting URL: https://www.youtube.com/watch?v=wsPub08_F4k
[youtube] wsPub08_F4k: Downloading webpage
WARNING: [youtube] Webpage contains broken formats (poToken experiment detected). Ignoring initial player response
[youtube] wsPub08_F4k: Downloading player f8071a08
[youtube] wsPub08_F4k: Downloading web player API JSON
WARNING: [youtube] API returned broken formats (poToken experiment detected). Retrying (1/3)...
[youtube] wsPub08_F4k: Downloading web player API JSON
WARNING: [youtube] API returned broken formats (poToken experiment detected). Retrying (2/3)...
[youtube] wsPub08_F4k: Downloading web player API JSON
WARNING: [youtube] API returned broken formats (poToken experiment detected). Retrying (3/3)...
[youtube] wsPub08_F4k: Downloading web player API JSON
WARNING: [youtube] API returned broken formats (poToken experiment detected). Giving up after 3 retries
ERROR: [youtube] wsPub08_F4k: Sign in to confirm you’re not a bot. This helps protect our community. Learn more
**

**Comment by bashonly** [comment]
@jzzbeyond you are missing a `;` before `player_skip=`

**Comment by jzzbeyond** [comment]
> @jzzbeyond you are missing a `;` before `player_skip=`
> 
> also it doesn't look like you're using the PR branch

thank you.  I updated the issue.  it still did not work.   I am not using the PR branch

**Comment by seproDev** [comment]
@jzzbeyond Please keep this PR for actual discussion of the code. This is not the place to ask tech support questions. If you are unfamiliar with how git works, please wait until this has been finalized and merged.

@Elite They seem to be valid for multiple requests and multiple days. I haven't tested the limits any further. I could imagine that these limits will also be adjusted over time.

**Comment by BigHeadDev** [comment]
> @jzzbeyond Please keep this PR for actual discussion of the code. This is not the place to ask tech support questions. If you are unfamiliar with how git works, please wait until this has been finalized and merged.
> 
> @Elite They seem to be valid for multiple requests and multiple days. I haven't tested the limits any further. I could imagine that these limits will also be adjusted over time.

May I ask if you can publish the release exe program in your repository for me to download and test?

**Comment by jzzbeyond** [comment]
> @jzzbeyond Please keep this PR for actual discussion of the code. This is not the place to ask tech support questions. If you are unfamiliar with how git works, please wait until this has been finalized and merged.
> 
> @Elite They seem to be valid for multiple requests and multiple days. I haven't tested the limits any further. I could imagine that these limits will also be adjusted over time.

sorry.  I am unfimiliar with that.  I will wait for the final result merged

File: README.md
Comment (coletdjnz) [inline README.md:1780]: maybe add a note that this extractor arg should not be used with cookies

Comment (coletdjnz) [inline README.md:1780]:
```suggestion
* `visitor_data`: Overwrites the visitor data used in Innertube API requests. Note this should only be used for testing without cookies. If a session is required you should be using cookies.
```

perhaps something like this?

File: README.md
Comment (coletdjnz) [inline README.md:1781]:
```suggestion
* `po_token`:  Proof of Origin (PO) token to use for all clients. This token is generated by BotGuard/DroidGuard to attest the requests are coming from a real client. Note: this needs to be paired with cookies from the browser session the token was generated from (or the `visitor_data` the token was generated with), and the matching `player_client`.
```

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:3793]:
```suggestion
```

The initial response may still work? If it's for a session generated from a browser user should be passing cookies, so the visitor ID will match token that way.

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:3797]:
```suggestion
                if not self.po_token and all(x in experiments for x in self._POTOKEN_EXPERIMENTS):
```

perhaps?

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:3793]: Tested and this works

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:652]: should be in _extract_player_response imo

Comment (bashonly) [reply yt_dlp/extractor/youtube.py:652]: I think this would work?
```diff
diff --git a/yt_dlp/extractor/youtube.py b/yt_dlp/extractor/youtube.py
index 224c9b988..6a4f7aadf 100644
--- a/yt_dlp/extractor/youtube.py
+++ b/yt_dlp/extractor/youtube.py
@@ -3734,6 +3734,9 @@ def _extract_player_response(self, client, video_id, master_ytcfg, player_ytcfg,
         yt_query = {
             'videoId': video_id,
         }
+
+        if self.po_token:
+            yt_query['serviceIntegrityDimensions'] = {'poToken': self.po_token}
 
         default_pp = traverse_obj(
             INNERTUBE_CLIENTS, (_split_innertube_client(client)[0], 'PLAYER_PARAMS', {str}))
```

**Comment by coletdjnz** [comment]
Wiki page to compliment this PR: https://github.com/yt-dlp/yt-dlp-wiki/pull/40

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:3842]: We can probably remove this initial logic here, since web will always require PO Token now

Keep it consistent with android client and mark the formats as broken unless PO token provided

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:3905]: Add a check that will only enter this branch if the client is not already marked as requires PO Token

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:748]: Pass data sync id too

Investigate whether session index is used

File: yt_dlp/extractor/youtube.py
Comment (seproDev) [reply yt_dlp/extractor/youtube.py:3842]: I think that's fair. I would assume the experiment IDs will also differ for other clients.

Comment (bashonly) [reply yt_dlp/extractor/youtube.py:3842]: the webpage data will always be `WEB` client right?

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:3816]: this doesn't do anything anymore; should we check if the client has `REQUIRE_PO_TOKEN` and append it to `broken_clients` if there is no accompanying `po_token` passed?

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:3850]: Yep

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:3816]: Hmm, we don't know about po_token at this point.. We would have to de-prioritize later 🤔

File: README.md
Comment (coletdjnz) [inline README.md:1782]: confirm syntax

**Review by seproDev — COMMENTED** [review]
LGTM (I can't approve since I opened the PR)

Thanks coletdjnz for doing a lot of the work here!

**Review by coletdjnz — APPROVED** [review]
lgtm

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:98]: `WEB_EMBEDDED_PLAYER` doesn't require a PO token?

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:706]:
```suggestion
        channel_syncid, _, user_syncid = data_sync_id.partition('||')
        if user_syncid:
            return channel_syncid
```

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:716]:
```suggestion
        if delegated_sid := traverse_obj(args, (..., 'DELEGATED_SESSION_ID', {str}, any)):
```

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:735]: I'm having a hard time following the traversal path with this formatting. Maybe could tighten it up by using `{str}` and `any`?

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [reply yt_dlp/extractor/youtube.py:3844]: Nothing addressed this, correct? Not saying we needed to, just confirming.

I do think we should remove this `broken_clients`/`BROKEN_CLIENTS` code since it is pointless and only making things more complicated than necessary

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:3935]:
```suggestion
                    f'--extractor-args "youtube:po_token={client}+XXX"',
```

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:3933]:
```suggestion
                    f'No PO Token provided for {client} client, '
                    f'which is required for working {client} formats. '
```

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:3950]:
```suggestion
                    po_token=po_token)
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/youtube.py
Comment (Grub4K) [reply yt_dlp/extractor/youtube.py:706]: Is `data_sync_id` always a string? if so, the `if not data_sync_id` could be omitted as well.

File: yt_dlp/extractor/youtube.py
Comment (Grub4K) [inline yt_dlp/extractor/youtube.py:3736]:
```suggestion
            po_token_client, sep, po_token = token_str.partition('+')
            if not sep:
                self.report_warning('Invalid po_token configuration format. Expected "client+po_token"')
                continue
            if not client or po_token_client == client:
                return po_token
```
maybe?

File: yt_dlp/extractor/youtube.py
Comment (Grub4K) [inline yt_dlp/extractor/youtube.py:3744]:
```suggestion
                'You may need to pass Visitor Data with --extractor-args youtube:visitor_data=XXX')
```
and same below?

**Review by Grub4K — APPROVED** [review]

File: yt_dlp/extractor/youtube.py
Comment (grqz) [reply yt_dlp/extractor/youtube.py:3744]: maybe also quote the extractor args

Comment (grqz) [reply yt_dlp/extractor/youtube.py:3744]:
```suggestion
                f'You may need to pass Visitor Data with --extractor-args "youtube:visitor_data=XXX"')
```

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:98]: Not currently last time I tested

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py]: updated. never used that syntax before, but looks like I got it working

File: yt_dlp/extractor/youtube.py
Comment (grqz) [inline yt_dlp/extractor/youtube.py:3748]:
```suggestion
                    f'You may need to pass a Data Sync ID with --extractor-args "youtube:data_sync_id=XXX"')
```

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:730]: need to move the `{str}` out of the branching tuple
```suggestion
        return traverse_obj(
            args, (..., ('DATASYNC_ID', ('responseContext', 'mainAppWebResponseContext', 'datasyncId')), {str}, any))
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:3727]: Could include the bad arg with an only_once
```suggestion
                self.report_warning(
                    f'Invalid po_token configuration format. Expected "client+po_token", got "{token_str}"', only_once=True)
```

File: yt_dlp/extractor/youtube.py
Comment (bashonly) [reply yt_dlp/extractor/youtube.py]: > Hmm, we don't know about po_token at this point..

can't we call `_get_config_po_token` here?

**Comment by bashonly** [comment]
for visibility, this is still unresolved: https://github.com/yt-dlp/yt-dlp/pull/10648#discussion_r1752654434

File: yt_dlp/extractor/youtube.py
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py]: No, as the po token may not come from the config, and we need the data sync ID/visitor data and other runtime info

**Comment by KennyTalarico** [comment]
Hi, thanks so much for all the work that goes into this project. I just wanted to make the developers aware that using PO token + visitor data fixed an IP ban for me (running on AWS), **but** this only lasted for about a week. Now, even with a refreshed PO token/visitor data (or cookies), including ones generated on someone else's machine not using the same network as me, I once again get the "sign in to confirm you're not a bot" error. Not asking for tech support, just flagging that it seems like with enough usage the IP ban will supersede a valid PO token.

**Review by bashonly — APPROVED** [review]

**Comment by bashonly** [comment]
This looks good to merge now IMO

**Pull request merged**
